### PR TITLE
feat: add simple tests for weights

### PR DESCRIPTION
**Pull request by drbh**
This PR adds a few more server tests for loading weights

### CODE REVIEW

File: server/tests/utils/test_weights.py [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+                [1, 2],
+                [3, 4],
+                [5, 6],
+                [7, 8],
+            ],
+            dtype=torch.float32,
+        ),
+    },
+    "test_get_multi_weights_row_gptq": {
+        "weight.qweight": torch.tensor(
```
Comment (danieldk) [inline server/tests/utils/test_weights.py:61]: I think the quantizers should probably be tested with the real data types to check that there are no accidental dtype conversions happening. So `int32` for `qweight`, `qzeros`, and `g_idx`, and `f16` for `scales`.

Maybe also real shapes? (e.g. currently `qzeros` does not line up with just having one group) Probably less important.

File: server/tests/utils/test_weights.py [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        ),
+        "weight.g_idx": torch.tensor([1.0], dtype=torch.float32),
+        "weight.qzeros": torch.tensor([[1.0], [2.0]], dtype=torch.float32),
+        "weight.scales": torch.tensor([8], dtype=torch.int32),
+        #
+        "gptq_bits": torch.tensor([8], dtype=torch.float32),
+        "gptq_groupsize": torch.tensor([4], dtype=torch.float32),
+    },
+    "test_get_multi_weights_row_exl2": {
+        "weight.q_weight": torch.tensor(
```
Comment (danieldk) [inline server/tests/utils/test_weights.py:78]: * `q_groups`: i16
* `q_weight`, `q_invperm`: i32
* `q_scale_max`: should be an f16 tensor

File: server/tests/utils/test_weights.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+            ],
+            dtype=torch.float32,
+        ),
+        "weight.q_scale": torch.tensor([8], dtype=torch.int32),
+        "weight.q_invperm": torch.tensor([1.0], dtype=torch.float32),
+        "weight.q_scale_max": 8,
+        "weight.q_groups": torch.tensor([4], dtype=torch.int32),
+    },
+    "test_get_multi_weights_row_marlin": {
+        "weight.scales": torch.tensor([8], dtype=torch.float16),
```
Comment (danieldk) [inline server/tests/utils/test_weights.py:93]:
```suggestion
```

`s` contains the scales

File: server/tests/utils/test_weights.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+
+    w = weights.get_weights_col_packed(
+        prefix=prefix,
+        quantize=quantize,
+        block_sizes=block_sizes,
+    )
+
+    prefix = "col_packed"
+    quantize = None
+    block_sizes = 1
```
Comment (danieldk) [inline server/tests/utils/test_weights.py:263]: Same as above?

Comment (danieldk) [inline server/tests/utils/test_weights.py:263]: Maybe it's also worth testing `block_sizes=2` and `block_sizes=[1,1]`?

File: server/tests/utils/test_weights.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+                [3, 4],
+                [5, 6],
+                [7, 8],
+            ],
+            dtype=torch.float32,
+        ),
+        "weight.g_idx": torch.tensor([1.0], dtype=torch.float32),
+        "weight.qzeros": torch.tensor([[1.0], [2.0]], dtype=torch.float32),
+        "weight.scales": torch.tensor([8], dtype=torch.int32),
+        #
```
Comment (danieldk) [inline server/tests/utils/test_weights.py:73]:
```suggestion
        #
```

Stray comment mark?

**Review by danieldk — COMMENTED** [review]
Really nice to have these tests!

Column gptq/marlin/exl2 and column-packed gptq/marlin tests seem to be missing?

**Comment by drbh** [comment]
updated to include more test (21)

```
python -m pytest -vvv server/tests/utils/test_weights.py

================================================================================== test session starts ===================================================================================
platform linux -- Python 3.11.1, pytest-7.4.0, pluggy-1.5.0 -- /home/ubuntu/Projects/text-generation-inference/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/ubuntu/Projects/text-generation-inference/server
configfile: pyproject.toml
plugins: asyncio-0.23.7, anyio-4.3.0, syrupy-4.0.1
asyncio: mode=Mode.STRICT
collected 21 items

server/tests/utils/test_weights.py::test_weights PASSED                                                                                                                            [  4%]
server/tests/utils/test_weights.py::test_get_tensor PASSED                                                                                                                         [  9%]
server/tests/utils/test_weights.py::test_get_weights_col_packed PASSED                                                                                                             [ 14%]
server/tests/utils/test_weights.py::test_get_weights_col_packed_block_size PASSED                                                                                                  [ 19%]
server/tests/utils/test_weights.py::test_get_weights_col_packed_block_size_arr PASSED                                                                                              [ 23%]
server/tests/utils/test_weights.py::test_get_multi_weights_col PASSED                                                                                                              [ 28%]
server/tests/utils/test_weights.py::test_get_multi_weights_row PASSED                                                                                                              [ 33%]
server/tests/utils/test_weights.py::test_get_weights_col_gtpq PASSED                                                                                                               [ 38%]
server/tests/utils/test_weights.py::test_get_weights_col_exl2 PASSED                                                                                                               [ 42%]
server/tests/utils/test_weights.py::test_get_weights_col_packed_awq PASSED                                                                                                         [ 47%]
server/tests/utils/test_weights.py::test_get_weights_col_packed_exl2 SKIPPED (Review expected functionality)                                                                       [ 52%]
server/tests/utils/test_weights.py::test_get_weights_col_packed_gptq PASSED                                                                                                        [ 57%]
server/tests/utils/test_weights.py::test_get_weights_col_packed_marlin PASSED                                                                                                      [ 61%]
server/tests/utils/test_weights.py::test_get_multi_weights_col_awq PASSED                                                                                                          [ 66%]
server/tests/utils/test_weights.py::test_get_multi_weights_col_exl2 PASSED                                                                                                         [ 71%]
server/tests/utils/test_weights.py::test_get_multi_weights_col_gptq PASSED                                                                                                         [ 76%]
server/tests/utils/test_weights.py::test_get_multi_weights_col_marlin PASSED                                                                                                       [ 80%]
server/tests/utils/test_weights.py::test_get_multi_weights_row_awq PASSED                                                                                                          [ 85%]
server/tests/utils/test_weights.py::test_get_multi_weights_row_exl2 PASSED                                                                                                         [ 90%]
server/tests/utils/test_weights.py::test_get_multi_weights_row_gptq PASSED                                                                                                         [ 95%]
server/tests/utils/test_weights.py::test_get_multi_weights_row_marlin PASSED                                                                                                       [100%]

-------------------------------------------------------------------------------- snapshot report summary ---------------------------------------------------------------------------------

============================================================================= 20 passed, 1 skipped in 3.67s ==============================================================================
```

File: server/tests/utils/test_weights.py (hunk 1 above)
Comment (drbh) [reply server/tests/utils/test_weights.py:61]: great points, I've updated the types in the latest changes but the values are still not fully realistic, happy to follow up in this or a future PR

File: server/tests/utils/test_weights.py (hunk 2 above)
Comment (drbh) [reply server/tests/utils/test_weights.py:109]: updated, thanks!

File: server/tests/utils/test_weights.py (hunk 3 above)
Comment (drbh) [reply server/tests/utils/test_weights.py]: removed

File: server/tests/utils/test_weights.py (hunk 4 above)
Comment (drbh) [reply server/tests/utils/test_weights.py:318]: removed

Comment (drbh) [reply server/tests/utils/test_weights.py:318]: added new tests for both

File: server/tests/utils/test_weights.py (hunk 5 above)
Comment (drbh) [reply server/tests/utils/test_weights.py]: oops thanks!

File: server/tests/utils/test_weights.py (hunk 1 above)
Comment (drbh) [reply server/tests/utils/test_weights.py:61]: also I'm not 100% sure if `col_packed_exl2` is suppose to be supported or not? I was a bit confused about the expected weight names ("weights" vs "qweights")... is this intended? I'll update the test accordingly 🙂

Comment (danieldk) [reply server/tests/utils/test_weights.py:73]: Great question! This path is missing in `weights`, I'll create a ticket and look it. I can update the test here once it's properly implemented.

**Review by danieldk — APPROVED** [review]

**Pull request merged**